core/src/main/java/org/elasticsearch/ResourceNotFoundException.java

```diff
@@ -0,0 +1,52 @@
+package org.elasticsearch;
+
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.index.Index;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.rest.RestStatus;
+
+import java.io.IOException;
+
+/**
+ * Generic ResourceNotFoundException corresponding to the {@link RestStatus#NOT_FOUND} status code
+ */
+public class ResourceNotFoundException extends ElasticsearchException {
+
+
+    public ResourceNotFoundException(ShardId shardId, String msg, Object... args) {
+        this(msg, args);
+        addShard(shardId);
+    }
+
+    public ResourceNotFoundException(ShardId shardId, String msg, Throwable cause, Object... args) {
+        this(msg, cause, args);
+        addShard(shardId);
+    }
+
+    public ResourceNotFoundException(String index, String msg, Object... args) {
```

Comment:
this is a dangerous constructor... consider:

```
new ResourceNotFoundException("could not find resource {}  due to {}", "r1", "whatever reason")
```
maybe make `addShard` and `addIndex` and remove these ctors?

```
public ResourceNotFoundException extends ElasticsearchException {


}
```